legate/core/machine.py

```diff
@@ -0,0 +1,274 @@
+        if kind not in self._proc_ranges:
+            return ProcessorRange.create(kind, 1, 1, 0)
+        return self._proc_ranges[kind]
+
+    def only(self, *kinds: ProcessorKind) -> Machine:
+        return Machine([self._get_range(kind) for kind in kinds])
+
+    def count(self, kind: ProcessorKind) -> int:
+        return len(self._get_range(kind))
+
+    def filter_ranges(
+        self, task_info: Any, variant_ids: dict[ProcessorKind, int]
+    ) -> Machine:
+        valid_kinds = tuple(
+            kind
+            for kind in self.kinds
+            if task_info.has_variant(variant_ids[kind])
+        )
+        if valid_kinds == self.kinds:
+            return self
+        else:
+            return self.only(*valid_kinds)
+
+    def __getitem__(self, key: Union[str, slice, int, ProcSlice]) -> Machine:
+        if isinstance(key, ProcessorKind):
+            return self.only(key)
+        elif isinstance(key, (slice, int)):
+            if len(self._proc_ranges.keys()) > 1:
+                raise ValueError(
+                    "Ambiguous slicing: slicing is not allowed on a machine "
+                    "with more than one processor kind"
+                )
+            k = key if isinstance(key, slice) else slice(key, key + 1)
+            return Machine([self.get_processor_range().slice(k)])
+        elif isinstance(key, tuple) and len(key) == 2:
+            kind, sl = key
+            if not isinstance(sl, slice):
+                raise KeyError(f"Invalid slicing key: {key}")
+            return Machine([self._get_range(kind).slice(sl)])
+        else:
+            raise KeyError(f"Invalid slicing key: {key}")
+
+    @staticmethod
+    def create_toplevel_machine(runtime: Runtime) -> Machine:
+        num_nodes = int(
+            runtime.core_context.get_tunable(
+                legion.LEGATE_CORE_TUNABLE_NUM_NODES,
+                ty.int32,
+            )
+        )
+
+        def create_range(kind: ProcessorKind) -> ProcessorRange:
+            tunable_name = f"LEGATE_CORE_TUNABLE_TOTAL_{kind.name}S"
+            tunable_id = getattr(legion, tunable_name)
+            num_procs = int(
+                runtime.core_context.get_tunable(tunable_id, ty.int32)
+            )
+            return ProcessorRange.create(
+                kind, num_procs // num_nodes, 0, num_procs - 1
+            )
+
+        result = Machine([create_range(kind) for kind in ProcessorKind])
+        if result.empty:
+            raise RuntimeError(
+                "No processors are available to run Legate tasks. Please "
+                "create at least one processor of any kind. "
+            )
+        return result
+
+    def __and__(self, other: Machine) -> Machine:
+        if self is other:
+            return self
+        result: list[ProcessorRange] = []
+        for kind, r in self._proc_ranges.items():
+            if kind not in other._proc_ranges:
+                continue
+            result.append(r & other._proc_ranges[kind])
```

Comment:
No need for negative condition and `continue` here
```suggestion
            if kind in other._proc_ranges:
                result.append(r & other._proc_ranges[kind])
```